File: servidor.js
const express = require("express");
const cors    = require("cors");
const fs      = require("fs");
const path    = require("path");
const { execSync } = require("child_process");
const { google } = require("googleapis");

const app  = express();
const PORT = process.env.PORT || 3000;

app.use(cors());
app.use(express.json({ limit: "50mb" }));

// ── Google Drive setup ────────────────────────────────────────────────────
function getDriveClient() {
  const creds = JSON.parse(process.env.GOOGLE_CREDENTIALS);
  const auth  = new google.auth.GoogleAuth({
    credentials: creds,
    scopes: ["https://www.googleapis.com/auth/drive.file"],
  });
  return google.drive({ version: "v3", auth });
}

async function subirDrive(drive, filePath, fileName, folderId, mimeType) {
  const res = await drive.files.create({
    requestBody: {
      name:    fileName,
      parents: [folderId],
    },
    media: {
      mimeType,
      body: fs.createReadStream(filePath),
    },
    fields: "id, webViewLink",
  });
  return res.data;
}

// ── Ruta principal: generar Excel + Word y subir a Drive ──────────────────
app.post("/generar", async (req, res) => {
  const datos = req.body;
  const ts    = Date.now();
  const tmpDir = `/tmp/mp_${ts}`;
  fs.mkdirSync(tmpDir, { recursive: true });

  try {
    // 1. Guardar datos JSON
    const jsonPath = path.join(tmpDir, "datos.json");
    fs.writeFileSync(jsonPath, JSON.stringify(datos, null, 2));

    const mes   = (datos.mes  || "MES").toUpperCase().replace(/\s+/g, "_");
    const anio  = datos.anio  || "2026";
    const slug  = (datos.edificio || "EDIFICIO").replace(/\s+/g,"_").replace(/[^A-Za-z0-9_]/g,"").slice(0,20);

    // 2. Generar Excel
    console.log("Generando Excel...");
    const excelOut = path.join(tmpDir, `PARAMETROS_${slug}_${mes}_${anio}.xlsx`);
    execSync(`python3 /app/generar_excel.py ${jsonPath} ${excelOut}`, { timeout: 60000 });

    // 3. Generar gráficos
    console.log("Generando gráficos...");
    const grafDir = path.join(tmpDir, "graficos");
    fs.mkdirSync(grafDir, { recursive: true });
    const grafResult = execSync(
      `python3 /app/generar_graficos.py ${grafDir} < ${jsonPath}`,
      { timeout: 60000, maxBuffer: 10 * 1024 * 1024 }
    );
    const grafPaths = JSON.parse(grafResult.toString().trim());

    // Agregar paths al JSON para el informe
    const datosConGraf = { ...datos, _graf_paths: grafPaths };
    fs.writeFileSync(jsonPath, JSON.stringify(datosConGraf, null, 2));

    // 4. Generar informe Word
    console.log("Generando informe Word...");
    const wordOut = path.join(tmpDir, `INFORME_${slug}_${mes}_${anio}.docx`);
    execSync(`node /app/generar_informe.js ${jsonPath} ${wordOut}`, { timeout: 120000 });

    // 5. Subir a Google Drive
    console.log("Subiendo a Drive...");
    const drive         = getDriveClient();
    const folderExcel   = process.env.DRIVE_FOLDER_EXCEL;
    const folderWord    = process.env.DRIVE_FOLDER_WORD;

    const [excelFile, wordFile] = await Promise.all([
      subirDrive(drive, excelOut, path.basename(excelOut), folderExcel,
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"),
      subirDrive(drive, wordOut,  path.basename(wordOut),  folderWord,
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document"),
    ]);

    console.log("✓ Archivos subidos a Drive");

    // 6. Limpiar temporales
    fs.rmSync(tmpDir, { recursive: true, force: true });

    res.json({
      ok:        true,
      excel_url: excelFile.webViewLink,
      word_url:  wordFile.webViewLink,
      excel_id:  excelFile.id,
      word_id:   wordFile.id,
    });

  } catch (err) {
    console.error("Error al generar:", err.message);
    fs.rmSync(tmpDir, { recursive: true, force: true });
    res.status(500).json({ ok: false, error: err.message });
  }
});

// ── Health check ──────────────────────────────────────────────────────────
app.get("/", (req, res) => {
  res.json({
    status:    "ok",
    app:       "MP Ingeniería — Servidor de parámetros",
    version:   "1.0.0",
    timestamp: new Date().toISOString(),
  });
});

app.listen(PORT, () => {
  console.log(`✓ Servidor MP Ingeniería corriendo en puerto ${PORT}`);
});

File: servidor.py
import os, json, tempfile, subprocess, shutil
import urllib.request, urllib.error
from flask import Flask, request, jsonify, make_response

app = Flask(__name__)
PORT = int(os.environ.get("PORT", 8080))

SUPABASE_URL = os.environ.get("SUPABASE_URL", "").rstrip("/")
SUPABASE_KEY = os.environ.get("SUPABASE_KEY", "")

# Valores por defecto si Supabase no responde
DEFAULT_CONFIG = {
    "tecnicos": [
        "Acevedo Donoso, Mauricio Eugenio",
        "Arango Bravo, Jaime Hernán",
        "Bahamondes Palome, Claudio Mario",
        "Bustillos Pradenas, Benjamín",
        "De la Cruz Carreño Miranda, Alfonso Albino",
        "González Conopoima, Efraín Manuel",
        "Hermosilla Vega, Ricardo Manuel",
        "Rivera Navarro, Ismael Luciano",
        "Sánchez Soto, César Fernando",
        "Torres Yáñez, Kevin Daniel",
        "Uribe López, Luis Mario"
    ],
    "jefes": [
        "Jorge Rivera Navarro",
        "Jose Perez Olave",
        "Nelson Mena Malet"
    ]
}

def sb_get(clave):
    """Lee un valor de la tabla config en Supabase."""
    try:
        url = f"{SUPABASE_URL}/rest/v1/config?clave=eq.{clave}&select=valor"
        req = urllib.request.Request(url, headers={
            "apikey": SUPABASE_KEY,
            "Authorization": f"Bearer {SUPABASE_KEY}"
        })
        with urllib.request.urlopen(req, timeout=5) as r:
            data = json.loads(r.read())
            if data:
                return data[0]["valor"]
    except Exception as e:
        print(f"Supabase GET error ({clave}):", e)
    return None

def sb_set(clave, valor):
    """Guarda un valor en la tabla config en Supabase (upsert)."""
    try:
        url = f"{SUPABASE_URL}/rest/v1/config"
        payload = json.dumps({"clave": clave, "valor": valor}).encode()
        req = urllib.request.Request(url, data=payload, method="POST", headers={
            "apikey": SUPABASE_KEY,
            "Authorization": f"Bearer {SUPABASE_KEY}",
            "Content-Type": "application/json",
            "Prefer": "resolution=merge-duplicates"
        })
        urllib.request.urlopen(req, timeout=5)
        return True
    except Exception as e:
        print(f"Supabase SET error ({clave}):", e)
    return False

@app.after_request
def add_cors(response):
    response.headers["Access-Control-Allow-Origin"]  = "*"
    response.headers["Access-Control-Allow-Methods"] = "GET, POST, OPTIONS, PUT, DELETE"
    response.headers["Access-Control-Allow-Headers"] = "Content-Type, Accept"
    return response

@app.errorhandler(Exception)
def handle_exception(e):
    import traceback
    print("ERROR:", traceback.format_exc())
    resp = jsonify({"ok": False, "error": str(e)})
    resp.status_code = 500
    resp.headers["Access-Control-Allow-Origin"] = "*"
    return resp

def get_drive_service():
    import google.oauth2.credentials
    import google.auth.transport.requests
    from googleapiclient.discovery import build
    creds = google.oauth2.credentials.Credentials(
        token=None,
        refresh_token=os.environ["GOOGLE_REFRESH_TOKEN"],
        client_id=os.environ["GOOGLE_CLIENT_ID"],
        client_secret=os.environ["GOOGLE_CLIENT_SECRET"],
        token_uri="https://oauth2.googleapis.com/token",
        scopes=["https://www.googleapis.com/auth/drive.file"],
        enable_reauth_refresh=False
    )
    request_session = google.auth.transport.requests.Request()
    creds.refresh(request_session)
    return build("drive", "v3", credentials=creds)

def subir_drive(service, file_path, file_name, folder_id, mime_type):
    from googleapiclient.http import MediaFileUpload
    meta  = {"name": file_name, "parents": [folder_id]}
    media = MediaFileUpload(file_path, mimetype=mime_type)
    f = service.files().create(body=meta, media_body=media, fields="id,webViewLink").execute()
    return f

# ── Endpoints de configuración global ──────────────────────────────────────

@app.route("/config", methods=["OPTIONS"])
def config_options():
    return make_response("", 200)

@app.route("/config", methods=["GET"])
def get_config():
    tecnicos = sb_get("tecnicos") or DEFAULT_CONFIG["tecnicos"]
    jefes    = sb_get("jefes")    or DEFAULT_CONFIG["jefes"]
    return jsonify({"ok": True, "tecnicos": tecnicos, "jefes": jefes})

@app.route("/config", methods=["POST"])
def save_config():
    datos = request.get_json()
    if "tecnicos" in datos:
        sb_set("tecnicos", datos["tecnicos"])
    if "jefes" in datos:
        sb_set("jefes", datos["jefes"])
    print(f"✓ Config guardada en Supabase")
    return jsonify({"ok": True})

# ── Endpoint principal ──────────────────────────────────────────────────────

@app.route("/", methods=["GET"])
def health():
    return jsonify({"status":"ok","app":"MP Ingeniería — Servidor de parámetros","version":"2.1.0"})

@app.route("/generar", methods=["OPTIONS"])
def generar_options():
    return make_response("", 200)

@app.route("/generar", methods=["POST"])
def generar():
    datos = request.get_json()
    tmp   = tempfile.mkdtemp()
    try:
        json_path = os.path.join(tmp, "datos.json")
        with open(json_path, "w") as f:
            json.dump(datos, f, ensure_ascii=False)

        mes  = (datos.get("mes") or "MES").upper().replace(" ", "_")
        anio = datos.get("anio") or "2026"
        slug = "".join(c if c.isalnum() or c=="_" else "_"
                       for c in (datos.get("edificio") or "EDIFICIO").replace(" ","_"))[:20]

        excel_out = os.path.join(tmp, f"PARAMETROS_{slug}_{mes}_{anio}.xlsx")
        subprocess.run(["python3", "/app/generar_excel.py", json_path, excel_out],
                       check=True, timeout=60)
        print(f"✓ Excel generado")

        graf_dir = os.path.join(tmp, "graficos")
        os.makedirs(graf_dir, exist_ok=True)
        gr = subprocess.run(["python3", "/app/generar_graficos.py", graf_dir],
            input=open(json_path,"rb").read(), capture_output=True, timeout=60)
        try: graf_paths = json.loads(gr.stdout.decode().strip())
        except: graf_paths = {}
        datos["_graf_paths"] = graf_paths
        with open(json_path, "w") as f:
            json.dump(datos, f, ensure_ascii=False)
        print(f"✓ Gráficos: {list(graf_paths.keys())}")

        word_out = os.path.join(tmp, f"INFORME_{slug}_{mes}_{anio}.docx")
        subprocess.run(["python3", "/app/generar_informe.py", json_path, word_out],
                       check=True, timeout=120)
        print(f"✓ Word generado")

        service      = get_drive_service()
        folder_excel = os.environ["DRIVE_FOLDER_EXCEL"]
        folder_word  = os.environ["DRIVE_FOLDER_WORD"]

        excel_file = subir_drive(service, excel_out, os.path.basename(excel_out),
            folder_excel, "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
        word_file  = subir_drive(service, word_out, os.path.basename(word_out),
            folder_word, "application/vnd.openxmlformats-officedocument.wordprocessingml.document")
        print(f"✓ Archivos subidos a Drive")

        return jsonify({"ok":True,
                        "excel_url":excel_file.get("webViewLink"),
                        "word_url": word_file.get("webViewLink")})

    except Exception as e:
        import traceback
        print("ERROR:", traceback.format_exc())
        return jsonify({"ok":False,"error":str(e)}), 500
    finally:
        shutil.rmtree(tmp, ignore_errors=True)

if __name__ == "__main__":
    print(f"✓ Servidor MP Ingeniería v2.1 corriendo en puerto {PORT}")
    app.run(host="0.0.0.0", port=PORT)
